src/zmalloc.c

```diff
@@ -67,10 +67,31 @@ void zlibc_free(void *ptr) {
 #define dallocx(ptr,flags) je_dallocx(ptr,flags)
 #endif
 
-#define update_zmalloc_stat_alloc(__n) atomicIncr(used_memory,(__n))
-#define update_zmalloc_stat_free(__n) atomicDecr(used_memory,(__n))
+#define MAX_THREADS 16 /* Keep it a power of 2 so we can use '&' instead of '%'. */
+#define THREAD_MASK (MAX_THREADS - 1)
 
-static redisAtomic size_t used_memory = 0;
+typedef struct used_memory_entry {
+    redisAtomic long long used_memory;
+    char padding[CACHE_LINE_SIZE - sizeof(long)];
+} used_memory_entry;
+
+static __attribute__((aligned(CACHE_LINE_SIZE))) used_memory_entry used_memory[MAX_THREADS];
+static redisAtomic size_t num_active_threads = 0;
+static __thread long my_thread_index = -1;
+
+void update_zmalloc_stat_alloc(long num) {
+    if (my_thread_index == -1) {
+        my_thread_index = atomicIncr(num_active_threads, 1) & THREAD_MASK;
```

Comment:
atomicIncr() returns the new value, which means the index 0 will never be used.
should use `atomicGetIncr()`?

Reply:
i'm wondering if we can make the main thread in an individual index?

Reply:
I believe this is practically this is what gone happened on most cases (up to 16 threads).